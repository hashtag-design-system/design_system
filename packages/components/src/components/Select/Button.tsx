import React, { useEffect } from "react";
import { useSelectContext } from "../../utils/contexts";
import { useClassnames, useDisabled } from "../../utils/hooks";
import { ButtonProps } from "../Button";
import { ComponentProps } from "../__helpers__";
import { DownArrowIcon } from "./__icons__/DownArrowIcon";

export type Props = {
  showValue?: boolean;
};

export type FProps = Props & ComponentProps<"summary", false> & Pick<ButtonProps, "state">;

export type SBProps = Props & Pick<FProps, "state" | "className">

export const Button: React.FC<FProps> = ({ state, className, showValue = true, children, ...props }) => {
  const [classNames, rest] = useClassnames<FProps>("select__btn btn btn-secondary", props, { stateToRemove: { state } });
  const { value, ref, setIsDisabled } = useSelectContext();
  const isDisabled = useDisabled<boolean>(props, state);

  useEffect(() => {
    setIsDisabled(isDisabled);
  }, [isDisabled, setIsDisabled])

  return (
    <summary ref={ref} className={classNames} data-testid="select-btn" {...rest}>
      <p style={{ width: rest.style?.width }}>{value && showValue ? value : children}</p>
      <DownArrowIcon />
    </summary>
  );
};

Button.displayName = "SelectButton";
