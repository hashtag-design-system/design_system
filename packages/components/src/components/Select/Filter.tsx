import { debounce } from "lodash";
import React, { useCallback, useState } from "react";
import { useSelectContext } from "../../utils/contexts";
import { useClassnames } from "../../utils/hooks";
import Input, { InputFProps } from "../Input";

// Big thanks to -> https://dev.to/_martinwheeler_/create-a-react-search-bar-that-highlights-your-results-4hdh

export type Props = {
  filterById?: boolean;
  bold?: boolean;
  debounceMs?: number;
};

export type FProps = Props & InputFProps;

export const Filter: React.FunctionComponent<FProps> = React.memo(
  ({ defaultValue, filterById, bold = true, state, debounceMs = 250, onChange, onKeyDown, ...props }) => {
    const [key, setKey] = useState<string | undefined>(undefined);
    const [classNames, rest] = useClassnames<FProps>("select__filter", props);

    const { items, setItems, setFilterValue, handleToggle } = useSelectContext();

    const handleSearch = debounce(
      (newVal: string[]) => {
        const suggestions = items.map(({ id, content, ...item }) => {
          let isShown: boolean = item.isShown;
          if (content) {
            isShown = newVal.every(val => content.toLowerCase().indexOf(val.toLowerCase()) > -1);
          }
          if (filterById && id) {
            isShown = newVal.some(val => id.toLowerCase().indexOf(val.toLowerCase()) > -1);
            // isShown = id.toLowerCase().indexOf(value.toLowerCase()) > -1;
          }
          return { ...item, id, content, isShown };
        });
        setItems(suggestions);
      },
      debounceMs,
      { leading: true }
    );

    const handleChange = useCallback(
      (e: React.ChangeEvent<HTMLInputElement>) => {
        const newVal = e.target.value.split(" ").filter(val => val.trim().length >= 1);

        handleSearch(newVal);
        setFilterValue(newVal);

        setKey(undefined);

        if (onChange) {
          onChange(e);
        }
      },
      [handleSearch, setFilterValue, onChange]
    );

    const handleKeyDown = useCallback(
      (e: React.KeyboardEvent<HTMLInputElement>) => {
        const key = e.code;
        setKey(key);

        switch (key) {
          case "Escape":
            if (handleToggle) {
              handleToggle(e);
            }
            e.currentTarget.blur();
            break;
          case "Enter":
            items
              .filter(item => item.isShown)
              .forEach((item, i) => {
                if (i === 0) {
                  if (item.ref && item.ref.current) {
                    item.ref.current.click();
                  }
                }
              });
        }

        if (onKeyDown) {
          onKeyDown(e);
        }
      },
      [items, handleToggle, onKeyDown]
    );

    return (
      <Input
        className={classNames}
        state={state ? state : key && key === "Tab" ? "focus" : "default"}
        onChange={e => handleChange(e)}
        onKeyDown={e => handleKeyDown(e)}
        data-testid="select-filter"
        {...rest}
      />
    );
  }
);
