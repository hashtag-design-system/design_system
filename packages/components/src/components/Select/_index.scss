@import "./__helpers__/mixins";

.select {
  &__container {
    @extend .flex-row-stretch-flex-start;
    z-index: 1;
  }

  &__btn {
    height: auto;
    justify-content: space-between;
    border-color: $input-border-color;
    border-width: $input-border-width / 2;
    box-shadow: 0px 0px 2px rgba(194, 194, 194, 0.25);

    @include hover-classNames() {
      background-color: $grey-2;
      color: initial;
      border-color: $input-border-color;
    }

    p {
      text-align: left;
      overflow: hidden;
      white-space: nowrap;
      text-overflow: ellipsis;
    }
  }

  &__options {
    max-height: em(480);
    height: auto;
    width: 100%;
    overflow-x: hidden;
    overflow-y: auto;

    &::-webkit-scrollbar {
      width: em(6);
      background-color: transparent;
      border-radius: inherit;
    }

    &::-webkit-scrollbar-thumb {
      background-color: $grey-6;
      background-clip: padding-box;
      border: 1px solid transparent;
      border-right-width: 2px;
      border-top-width: 4px;
      border-bottom-width: 4px;
      border-radius: inherit;
      box-shadow: none;
    }
  }

  &__modal {
    position: absolute;
    inset: auto;
    padding: 0;
    margin: em(8) 0 em(0) 0;
    background-color: $grey-1;
    z-index: 99;
    border-radius: 6px;
    pointer-events: auto;
    border: $input-border-width / 2 solid $input-border-color;
    height: auto;
    width: auto;
    box-sizing: border-box;
    overflow: hidden;
    z-index: 99;
    box-shadow: 0px 1px 4px rgba(194, 194, 194, 0.5);

    &--mobile {
      @extend .flex-row-center-center;
    }
  }

  &__item {
    padding: em(14);
    border-radius: 0;
    user-select: none;
    z-index: 1;

    @include hover-classNames() {
      @include breakpoint("sm") {
        @include select__item--hover();
      }
    }

    @include focus-classNames() {
      @extend .select__item--hover;
    }

    &[aria-selected="true"] {
      @extend .select__item--hover;
    }

    label {
      cursor: pointer;
      user-select: none;
      color: $grey-9;
    }

    input {
      display: none;
    }

    + .select__hr {
      margin-top: 0;
      margin-bottom: 0;
    }
  }

  &__header {
    @extend .select__item;
    border-radius: 0;
    padding-top: em(8);
    padding-bottom: em(8);

    &[data-children="true"] {
      & ~ .select__item:not(.select__header),
      & ~ .select__options .select__item:not(.select__header) {
        margin-left: em(20);
      }
    }

    h6 {
      font-size: em(12);
    }

    + .select__hr {
      margin-top: 0;
    }
  }

  &__hr {
    display: block;
    width: auto;
    margin: em(4) em(0);
    height: $input-border-width / 2;
    background: $input-border-color;
  }

  &__icon {
    fill: none !important;
    stroke: $grey-7 !important;
  }
}

@include breakpoint("sm", "min-width") {
  .select {
    &__item {
      padding: em(6) em(4);
      margin: em(4);
      border-radius: em(4);
    }

    &__header {
      padding-top: em(2);
      padding-bottom: em(2);
    }

    &__hr {
      margin: em(4);
    }

    &__options {
      max-height: em(200);
    }
  }
}

@include modal-classNames("select__modal--mobile") {
  .select__modal {
    width: min(70%, 32em);
    max-height: 80%;
  }

  .select__item:not(.select__header) {
    &:last-of-type {
      + .select__hr {
        display: none;
      }
    }
  }
}
