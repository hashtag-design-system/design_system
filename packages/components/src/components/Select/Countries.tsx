import { COUNTRIES } from "@georgekrax-hashtag/common";
import React, { useMemo } from "react";
import { useClassnames } from "../../utils/hooks";
import Select, { SelectItemFProps } from "./index";

export const countriesArr = Object.values(COUNTRIES);
export type CountriesArrType = typeof countriesArr;

export type Props = {
  withFlags?: boolean;
  arr?: (countries: CountriesArrType) => CountriesArrType;
};

export type FProps = Props & Partial<SelectItemFProps>;

const Countries: React.FC<FProps> = React.memo(({ withFlags = true, arr, ...props }) => {
  const [classNames, rest] = useClassnames("select__item__country", props);

  // Works the same, if useEffect() and useState() were used, on initial render
  const fArr = useMemo(() => (arr ? arr(countriesArr) : countriesArr), [arr]);

  return (
    <>
      {fArr.map(({ id, name, flagKey }) => {
        const imageRequest = JSON.stringify({
          bucket: "hashtag.countries",
          key: `flags/${flagKey}`,
          edits: {
            resize: {
              width: 22,
              height: 16,
              fit: "fill",
            },
          },
        });
        const url = `https://d32454kkzii6gk.cloudfront.net/${btoa(imageRequest)}`;
        // const url = require(`../../../../assets/${flagKey}`)["default"];

        return (
          <Select.Item
            key={id}
            id={name.toLowerCase()}
            content={name}
            className={classNames}
            htmlContent={
              withFlags
                ? {
                    before: <img src={url} className="select__item__country-flag" alt={name.toLowerCase() + "_flag"} />,
                  }
                : undefined
            }
            {...rest}
          />
        );
      })}
    </>
  );
});

Countries.displayName = "SelectCountries";

export default Countries;
