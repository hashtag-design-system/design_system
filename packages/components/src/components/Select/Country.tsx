import { COUNTRY_TYPE } from "@georgekrax-hashtag/common";
import React, { useEffect, useState } from "react";
import { useClassnames } from "../../utils/hooks";
import Select, { SelectCountriesProps, SelectItemFProps } from "./index";

export type FProps = COUNTRY_TYPE & Pick<SelectCountriesProps, "withFlags"> & Partial<Omit<SelectItemFProps, "id">>;

const Country: React.FC<FProps> = ({
  id,
  name,
  flagKey,
  alpha2Code,
  alpha3Code,
  callingCode,
  isEu,
  withFlags = true,
  ...props
}) => {
  const [url, setUrl] = useState<string>("");
  const [classNames, rest] = useClassnames("select__item__country", props);

  useEffect(() => {
    const imageRequest = JSON.stringify({
      bucket: "hashtag.countries",
      key: `flags/${flagKey}`,
      edits: {
        resize: {
          width: 22,
          height: 16,
          fit: "fill",
        },
      },
    });
    const newUrl = `https://d32454kkzii6gk.cloudfront.net/${btoa(imageRequest)}`;
    setUrl(newUrl);
  }, [flagKey]);

  return (
    <Select.Item
      id={name.toLowerCase()}
      content={name}
      className={classNames}
      htmlContent={
        withFlags
          ? {
              before: <img src={url} className="select__item__country-flag" alt={name + " flag"} />,
            }
          : undefined
      }
      {...rest}
    />
  );
};

Country.displayName = "SelectCountry";

export default Country;