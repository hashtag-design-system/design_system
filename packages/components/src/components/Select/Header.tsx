import React from "react";
import { useClassnames } from "../../utils/hooks";
import { ComponentProps } from "../__helpers__";
import Select from "./index";

export type Props = {
  value: React.ReactNode;
};

export type FProps = Props & ComponentProps<"div">;

export const Header: React.FunctionComponent<FProps> = ({ value, children, ...props }) => {
  const [classNames, rest] = useClassnames<Partial<FProps>>("select__header", props);

  return (
    <>
      <div className={classNames} data-testid="select-header" data-children={children ? true : false} {...rest}>
        <h6>{value}</h6>
      </div>
      <Select.Hr />
      {children}
    </>
  );
};

Header.displayName = "SelectHeader";
