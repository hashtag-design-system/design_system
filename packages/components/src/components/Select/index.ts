import Select from "./Select";

export type { FProps as SelectButtonFProps, Props as SelectButtonProps, SBProps as SelectButtonSBProps } from "./Button";
export type { FProps as SelectFilterFProps, Props as SelectFilterProps } from "./Filter";
export type { FProps as SelectHeaderFProps, Props as SelectHeaderProps } from "./Header";
export type { FProps as SelectHrFProps } from "./Hr";
export type { Props as SelectItemProps, FProps as SelectItemFProps, SBProps as SelectItemSBProps } from "./Item";
export { SelectModalAligns } from "./Modal";
export type { FProps as SelectModalFProps, Props as SelectModalProps } from "./Modal";
export type { FProps as SelectOptionsFProps } from "./Options";
export type { FProps as SelectFProps, Props as SelectProps, SBProps as SelectSBProps } from "./Select";
export type { FProps as SelectCountriesFProps, Props as SelectCountriesProps } from "./Countries";

export default Select;
