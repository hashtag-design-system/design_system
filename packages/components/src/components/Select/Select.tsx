import React, { useEffect, useRef, useState } from "react";
import { SelectContextProvider } from "../../utils/contexts";
import { useClassnames, useClickOutside, useIsMobile } from "../../utils/hooks";
import { ComponentProps } from "../__helpers__";
import { Button } from "./Button";
import { Filter } from "./Filter";
import { Header } from "./Header";
import { Hr } from "./Hr";
import { Item } from "./Item";
import { Modal } from "./Modal";
import { Options } from "./Options";

export type SelectedItems = { id: string; content: string; highlightedChildren: string; selected: boolean; isShown: boolean };

export type Props = {
  defaultOpen?: boolean;
  multiSelectable?: boolean;
  mobileView?: boolean;
  onSelect?: (selected: SelectedItems[]) => void;
  onDismiss?: (e: React.MouseEvent<HTMLElement>) => void;
  onOutsideClick?: (outsideClick: boolean) => void;
};

type SubComponents = {
  Item: typeof Item;
  Header: typeof Header;
  Modal: typeof Modal;
  Button: typeof Button;
  Hr: typeof Hr;
  Options: typeof Options;
  Filter: typeof Filter;
};

export type FProps = Props &
  Omit<ComponentProps<"details", false>, "onSelect"> & {
    forwardRef?: ComponentProps<"details", true>["ref"];
  };

export type SBProps = Props & Pick<FProps, "defaultOpen" | "onToggle" | "onSelect">;

const Select: React.FC<FProps> & SubComponents = ({
  defaultOpen = false,
  multiSelectable = false,
  mobileView = false,
  children,
  open,
  forwardRef,
  onToggle,
  onSelect,
  onOutsideClick,
  onDismiss,
  ...props
}) => {
  const { ref: modalRef, isOpen, setIsOpen, outsideClick } = useClickOutside<HTMLDivElement>(defaultOpen, undefined, onDismiss);
  const [value, setValue] = useState<string>("");
  const [items, setItems] = useState<SelectedItems[]>([]);
  const [isDisabled, setIsDisabled] = useState(false);
  const [classNames, rest] = useClassnames("select__box__container", props);
  const { isMobile } = useIsMobile(mobileView);

  useEffect(() => {
    if (onOutsideClick) {
      onOutsideClick(outsideClick);
    }
  }, [outsideClick, onOutsideClick]);

  const divRef = useRef<HTMLDivElement>(null);

  const handleClick = (e: React.MouseEvent<HTMLElement>) => {
    if (isDisabled) {
      e.preventDefault();
    }
  };

  const handleToggle = (e: React.SyntheticEvent<HTMLElement>, boolean = true) => {
    e.preventDefault();

    if (!isDisabled) {
      if (boolean) {
        // @ts-expect-error
        setIsOpen(e.target.open);
      } else {
        setIsOpen(!isOpen);
      }
    }

    if (onToggle) {
      onToggle(e);
    }
  };

  const handleKeyDown = (e: React.KeyboardEvent<HTMLElement>) => {
    if (e.target instanceof HTMLInputElement) {
      return;
    }
    e.preventDefault();

    const focusItem = (next: boolean) => {
      if (divRef && divRef.current) {
        const options = Array.from(divRef.current.getElementsByClassName("select__item"));
        const selected = document.activeElement;

        let item: Element = options[0];
        if (selected && selected.tagName.toLowerCase() === "div") {
          const idx = options.indexOf(selected);
          if (next) {
            const nextOpt = options[idx + 1];
            if (nextOpt) {
              item = nextOpt;
              // (nextOpt as HTMLDivElement).focus();
            }
          } else {
            const nextOpt = options[idx - 1];
            if (nextOpt) {
              item = nextOpt;
              // (nextOpt as HTMLDivElement).focus();
            } else {
              item = options[options.length - 1];
            }
          }
        }
        return item;
      }
    };

    const isSummaryFocused = e.target instanceof Element && e.target.tagName.toLowerCase() === "summary";
    switch (e.code) {
      case "Escape":
        if (isOpen) {
          handleToggle(e, false);
        }
        break;
      case "Space":
      case "Enter":
        handleToggle(e, false);

        const selected = document.activeElement;
        if (!isSummaryFocused) {
          if (selected && selected.className === "select__item") {
            // @ts-expect-error
            document.activeElement.click();
            e.preventDefault();
          }
        }

        break;
      case "ArrowDown":
        if (isSummaryFocused && !isOpen) {
          handleToggle(e, false);
        } else {
          const target = focusItem(true);
          if (target) {
            (target as HTMLDivElement).focus();
          }
          e.preventDefault();
        }
        break;
      case "ArrowUp":
        if (isSummaryFocused && isOpen) {
          handleToggle(e, false);
        } else {
          const target = focusItem(false);
          if (target) {
            (target as HTMLDivElement).focus();
          }
          e.preventDefault();
        }
        break;
    }
  };

  useEffect(() => {
    setValue(
      items
        .filter(item => item.selected)
        .map(item => item.content)
        .join(", ")
    );

    if (onSelect && items.length >= 1) {
      onSelect(items);
    }
  }, [items, onSelect]);

  const fOpen = open === undefined ? isOpen : open;

  return (
    <SelectContextProvider
      value={{
        isOpen: fOpen,
        ref: forwardRef,
        value,
        multiSelectable,
        items,
        isMobile,
        modalRef,
        setItems,
        handleToggle,
        setIsDisabled,
      }}
    >
      <div className="select__container" ref={divRef} data-testid="select-container">
        <details
          ref={isMobile ? undefined : modalRef}
          className={classNames}
          open={fOpen}
          onClick={e => handleClick(e)}
          onToggle={e => handleToggle(e)}
          onKeyDown={e => handleKeyDown(e)}
          data-testid="select"
          {...rest}
        >
          {children}
        </details>
      </div>
    </SelectContextProvider>
  );
};

Select.Header = Header;
Select.Item = Item;
Select.Modal = Modal;
Select.Button = Button;
Select.Hr = Hr;
Select.Options = Options;
Select.Filter = Filter;

export default Select;
