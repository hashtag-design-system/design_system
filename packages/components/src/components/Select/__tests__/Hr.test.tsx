import { render, screen } from "@testing-library/react";
import Select from "../index";

describe("<Select.Hr />", () => {
  test("default behaviour", () => {
    render(<Select.Hr />);
    const hr = screen.getByTestId("select-hr");

    expect(hr).toBeVisible();
    expect(hr.tagName.toLowerCase()).toBe("span");
    expect(hr).toHaveAttribute("class");
    expect(hr.children).toHaveLength(0);
  });
  test('with className="custom-className"', () => {
    render(<Select.Hr className="custom-className" />);
    const hr = screen.getByTestId("select-hr");

    expect(hr).toHaveClass("custom-className");
    expect(hr).toMatchSnapshot();
  });
});
