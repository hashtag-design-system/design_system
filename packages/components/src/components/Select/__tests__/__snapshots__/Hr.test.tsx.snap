// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select.Hr /> with className="custom-className" 1`] = `
<span
  class="select__hr custom-className"
  data-testid="select-hr"
/>
`;
